### ISSUE TITLE
SPDK bdevperf crashes on segmentation fault when spdk_top is run in parallel

### ISSUE THREAD

**Issue by sziviziv** [spdk/spdk#2384]
Hi,

I am running bdevperf vs. NULL BDEV with socket opened in /var/tmp/spdk.sock. While test is running, I execute ./build/bin/spdk_top. 
spdk_top GUI opens and immediately closes. bdevperf test stops with a print "segmentation fault"

Expected Behavior:
spdk_top should run in parallel to bdevperf test

Steps to Reproduce:
In one window run: 
sudo ./test/bdev/bdevperf/bdevperf -c null_bdev.conf -q 256 -o 4096 -w randread -t 100 -m 0x1 -C -r /var/tmp/spdk.sock
in another window run 
sudo ./build/bin/spdk_top

Current print of test failure after spdk_top is executed:
sudo ./test/bdev/bdevperf/bdevperf -c null_bdev.conf -q 256 -o 4096 -w randread -t 100 -m 0x1 -r /var/tmp/spdk.sock

[2022-02-20 17:11:14.970647] Starting SPDK v22.05-pre git sha1 c833f6aa3 / DPDK 21.11.0 initialization...
[2022-02-20 17:11:14.970675] [ DPDK EAL parameters: [2022-02-20 17:11:14.970681] bdevperf [2022-02-20 17:11:14.970684] --no-shconf [2022-02-20 17:11:14.970687] -c 0x1 [2022-02-20 17:11:14.970689] --huge-unlink [2022-02-20 17:11:14.970692] --log-level=lib.eal:6 [2022-02-20 17:11:14.970695] --log-level=lib.cryptodev:5 [2022-02-20 17:11:14.970697] --log-level=user1:6 [2022-02-20 17:11:14.970700] --iova-mode=pa [2022-02-20 17:11:14.970703] --base-virtaddr=0x200000000000 [2022-02-20 17:11:14.970705] --match-allocations [2022-02-20 17:11:14.970708] --file-prefix=spdk_pid21385 [2022-02-20 17:11:14.970711] ]
EAL: No available 1048576 kB hugepages reported
TELEMETRY: No legacy callbacks, legacy socket not created
[2022-02-20 17:11:14.979581] app.c: 601:spdk_app_start: *NOTICE*: Total cores available: 1
[2022-02-20 17:11:15.065632] reactor.c: 943:reactor_run: *NOTICE*: Reactor started on core 0
[2022-02-20 17:11:15.065884] accel_engine.c: 510:spdk_accel_engine_initialize: *NOTICE*: Accel engine initialized to use software engine.
Running I/O for 100 seconds...
**Segmentation fault**

Thanks,

Ziv.

**Comment by yidong0635** [comment]
issue is reproduced. 
and the stack is:
```
Thread 1 "reactor_0" hit Breakpoint 3, rpc_framework_get_scheduler (request=0xb621e0, params=0x0) at app_rpc.c:519
519             struct spdk_scheduler *scheduler = spdk_scheduler_get();
(gdb) s
spdk_scheduler_get () at reactor.c:141
141             return g_scheduler;
(gdb)
rpc_framework_get_scheduler (request=0xb621e0, params=0x0) at app_rpc.c:520
520             uint64_t scheduler_period = spdk_scheduler_get_period();
(gdb)
spdk_scheduler_get_period () at reactor.c:148
148             return (g_scheduler_period * SPDK_SEC_TO_USEC / spdk_get_ticks_hz());
(gdb)
spdk_get_ticks_hz () at env.c:340
340             return rte_get_timer_hz();
(gdb)
rte_get_timer_hz () at /home/yidong/spdk/dpdk/build/include/generic/rte_cycles.h:122
122                     return rte_get_tsc_hz();
(gdb)

Thread 1 "reactor_0" received signal SIGSEGV, Segmentation fault.
rpc_framework_get_scheduler (request=0xb621e0, params=0x0) at app_rpc.c:531
531             spdk_json_write_named_string(w, "scheduler_name", scheduler->name);
(gdb) p scheduler
$1 = (struct spdk_scheduler *) 0x0

```

**Comment by yidong0635** [comment]
Maybe this is ok. It's good for me.  https://review.spdk.io/gerrit/c/spdk/spdk/+/11677

**Comment by sziviziv** [comment]
Thanks.
When I run spdk_top, I see that GUI shows a message at the bottom of the screen:
"ERROR occurred while getting scheduler data". 
Not sure if it affects the correctness of the displayed data.

**Comment by yidong0635** [comment]
@sziviziv  this should be another issue. And @KKaras169169 has some works about this。

**Comment by jimharris** [comment]
@sziviziv, that error does not affect the correctness of the displayed data.  You can ignore that for now.

@KKaras169169 - maybe for now, g_scheduler_data should return 0 instead of -EINVAL if we can call the framework_get_scheduler RPC but the results don't contain the scheduler_name?  This will be the case for applications like bdevperf that don't use SPDK reactors (and hence don't set a scheduler).

Currently that scheduler data isn't consumed in spdk_top, so once it is, it will need to account for the scheduler_name not being set.

**Comment by jimharris** [comment]
> Maybe this is ok. It's good for me. https://review.spdk.io/gerrit/c/spdk/spdk/+/11677

Patch merged.  Keeping issue open for now, but assigning to @KKaras169169 while we decide how to handle error messages on spdk_top side.

**Comment by KKaras169169** [comment]
spdk_top sends RPC request to get scheduler data from SPDK target (in this case bdevperf), which breaks the application, as it does not expect to send this data.

I will make sure scheduler is not NULL in rpc_framework_get_scheduler() function and modify spdk_top to take an argument to ignore scheduler data for cases such as this one (otherwise an error is printed in the bottom part of the window each time we request scheduler data via RPC - that may be confusing to some people).

**Comment by KKaras169169** [comment]
Here is the patch: https://review.spdk.io/gerrit/c/spdk/spdk/+/11781

**Comment by jimharris** [comment]
> Here is the patch for spdk_top side: https://review.spdk.io/gerrit/c/spdk/spdk/+/11781

Thanks @KKaras169169!  I just realized there might be a simpler way to fix this, than what I originally proposed.  Can you take a look at the comment on Gerrit?

**Comment by KKaras169169** [comment]
> > Here is the patch for spdk_top side: https://review.spdk.io/gerrit/c/spdk/spdk/+/11781
> 
> Thanks @KKaras169169! I just realized there might be a simpler way to fix this, than what I originally proposed. Can you take a look at the comment on Gerrit?

Done

**Comment by tomzawadzki** [comment]
Patch https://review.spdk.io/gerrit/c/spdk/spdk/+/11781 fixing the issue is now merged.
